Docs/Getting started 9c with the Unity editor (#19)
planetarium/ninechronicles-devforum: docs/getting-started-9c-with-the-unity-editor/getting-started-9c-with-the-unity-editor.md

---
title: Getting started Nine Chronicles with the Unity editor
---

> The [Nine Chronicles][9c-page] which is published by [Planetarium][planetarium-page] is developed with [Unity][unity-page]. It means the Nine Chronicles can execute by Unity Editor, And this article will show you how.

Reviewer: ```suggestion
The [Nine Chronicles][9c-page] published by [Planetarium][planetarium-page] is developed with [Unity][unity-page]. It means the Nine Chronicles can be executed by Unity Editor, and this article will show you how.
```

# Install Unity

Here's how to install Unity with Unity Hub.

## Install Unity Hub

### 1. Click the Unity Hub download link for your operating system.

- [Windows(.exe)](https://public-cdn.cloud.unity3d.com/hub/prod/UnityHubSetup.exe)
- [Mac(.dmg)](https://public-cdn.cloud.unity3d.com/hub/prod/UnityHubSetup.dmg)
- If the above link does not work, download Unity Hub from [Unity download page][unity-download-page].

### 2. Run the downloaded Unity Hub installation file.

[unity-download-page]: https://unity3d.com/get-unity/download

## Install Unity(2021.3.5f1)

### 1. Click [Unity(2021.3.5f1) installation link for Unity Hub][unity-install-with-hub-2021.3.5f1].

- If the above link does not work, download Unity(2021.3.5f1) from [Unity downlad archive page][unity-download-archive-page]. At this time, install with the green Unity Hub button.

  <img
    alt="Web: green Unity Hub button"
    src="./images/web-green-unity-hub-button.png"
    style="width:50vw" />

### 2. Unity(2021.3.5f1) installation proceeds with Unity Hub running.

[unity-install-with-hub-2021.3.5f1]: unityhub://2021.3.5f1/40eb3a945986
[unity-download-archive-page]: https://unity3d.com/get-unity/download/archive

> # Cloning the Nine Chronicles repository

Reviewer: ```suggestion
# Prepare the Nine Chronicles repository
```

Here I cover how to use [Git CLI][git-page] and [Fork][fork-page] as a way to clone a repository.

[git-page]: https://git-scm.com/
[fork-page]: https://git-fork.com/

## Cloning with the Git CLI

### 1. Install the Git CLI for your operating system.

- [Windows](https://git-scm.com/download/windows)
- [Mac](https://git-scm.com/download/mac)
- [Linux](https://git-scm.com/download/linux)

### 2. Clone the Nine Chronicles repository.

```
$ git clone https://github.com/planetarium/NineChronicles.git
$ cd NineChronicles
$ git submodule update --init --recursive
$ git config core.hooksPath hooks
```

### 3. Checkout to the `main` branch.

```
$ git checkout main
$ git submodule update --recursive

> ```

Reviewer: How about `development` branch? It would be very nice to add when to use `development` branch as comment or remark.

Author: @U-lis This article explains how to play by connecting to the main network with the Unity editor. So we need to checkout to the `main` branch.

## Cloning with the Fork

### 1. Install the Fork for your operating system.

- [Windows(.exe)](https://git-fork.com/update/win/ForkInstaller.exe)
- [Mac(2.20.2)(.dmg)](https://cdn.fork.dev/mac/Fork-2.20.2.dmg)
- If above link does not work or if you want to install another version, go to the [Fork page][fork-page] and download it.

### 2. Run the Fork and clone the Nine Chronicles repository.

- Click `File > Clone` menu.

  <img
    alt="Fork: `File > Clone` menu"
    src="./images/fork-menu-file-clone.png"
    style="width:50vw" />

- Enter `URL` and `Location` ant click the `Clone` button.
  - URL: `https://github.com/planetarium/NineChronicles.git`

>   - Location: In the image below, I spesified it as "`/Users/seungmin/Repositories/NineChronicles`".

Reviewer: ```suggestion
- Enter `URL` and `Location` and click the `Clone` button.
  - URL: `https://github.com/planetarium/NineChronicles.git`
  - Location: In the image below, I specified it as "`/Users/seungmin/Repositories/NineChronicles`".
```

    src="./images/fork-checkout-main.png"
    style="width:50vw" />

- In the following popup, click the `Track` button.

  <img
    alt="Fork: `Track` popup"
    src="./images/fork-track-popup.png"
    style="width:50vw" />

# Play the Nine Chronicles with the Unity editor

## Open the Nine Chronicles project with the Unity Hub

### 1. Run the Unity Hub.

Unity account creation and licensing are not covered here. You can also proceed for free(Unity Personal License).

### 2. Open the Nine Chronicles project.

- Click the `Projects > Open` button when running for the first time.

  <img
    alt="Unity Hub: `Projects > Open` button"
    src="./images/unity-hub-projects-open.png"
    style="width:50vw" />

- Select the `/nekoyume` folder in the location where you cloned the Nine Chronicles repository and click the `Open` button.

  <img
    alt="Unity Hub: `Open Project` popup"
    src="./images/unity-hub-projects-open-popup.png"
    style="width:50vw" />

## Configure `clo.json`

The clo.json file is a file that sets various options related to the blockchain network.

### 1. Create a clo.json file in `/nekoyume/Assets/StreamingAssets/` based on the cloned path.

> ### 2. Fill in the clo.json file as shown below and save it.

Reviewer: I think we should comment that the following sample `clo.json` is settings to attach mainnet in this section.
And we can suggest or provide settings for local blockchain network or something.

Author: @U-lis The purpose of this article is to connect to the main network and play with the Unity editor. I will change the title of the article and the first description.

```json
{
    "GenesisBlockPath": "https://download.nine-chronicles.com/genesis-block-Nine Chronicles-main",
    "NoMiner": true,
    "RpcClient": true,
    "RpcServerHost": "9c-main-rpc-1.nine-chronicles.com",
    "RpcServerPort": 31238,
    "ApiServerHost": "https://api.9c.gg/graphql"
}
```

## Prepare the key in the `Key Store` path

In order to play the Nine Chronicles, you need a Private Key to participate in the blockchain network.
If you've already played the Nine Chronicles, that means you already have a Private Key and a Protected Private Key.

- Private Key: A key required to sign a transaction.

> - Protected Private Key: This is the encrypted file of the private key.

Reviewer: ```suggestion
- Protected Private Key: An encrypted file of the private key.
```

### 1. Generate Private Key

> - Skip this step and go to the [Play](#play) step to use the step to generate a private key.

Reviewer: I think this sentence means someone who already has private key does't need to create private key but just use existing one.
If it's right, you should move this sentence to heading 2 section and show how to locate `key store path`.
It's awkward to notice of skipping whole process inside substep.

Author: @U-lis OK I'll rewrite it.

### 2. Put your Protected Private Key into the Key Store path

- Key Store path
  - The path to keep the Protected Private Key.
  - The default value of this path is determined [this way] [default-path-of-key-store] based on [Libplanet 0.41.0 version][libplanet-0.41.0].
  - Windows
    - `%appdata%\planetarium\keystore\`
  - Mac
    - `/Users/USERNAME/.config/planetarium/keystore/` or
    - `/Users/USERNAME/Library/Application Support/planetarium/keystore/`
  - Linux

>     - `home/planetarium/keystore/` or
>     - `home/.config/planetarium/keystore/`

Reviewer: For linux, each user's home directory is shortened as `~`.
```suggestion
    - `~/planetarium/keystore/` or
    - `~/.config/planetarium/keystore/`
```

[default-path-of-key-store]: https://github.com/planetarium/libplanet/blob/0.41.0/Libplanet/KeyStore/Web3KeyStore.cs#L21-L29

> [libplanet-0.41.0]: https://github.com/planetarium/libplanet/tree/0.41.0

Reviewer: Do we have to specify version? It's to hard to edit version everytime libplanet has bumped.

Author: @U-lis The logic to determine the Key Store path is in libplanet. I wrote the description based on the most recent version, and if this logic changes later, I thought to change the libplanet version and description together.
If we link the `main` branch on behalf of a specific version, there might be differences in the description and the logic the link points to, and I'd rather avoid this.

### 3. (not recommended) Setting the private key directly in the clo.json file

- Add a `"PrivateKey"` entry to the existing clo.json file.

  ```json
  {
      "PrivateKey": "your-pivate-key"
  }
  ```

## Play

> You are now ready to access the Nine Chronicles mainnet and play.

Reviewer: ```suggestion
You are now ready to access to the Nine Chronicles mainnet and play.
```

### 1. Open the `Game` scene.

> - Double-click the Game scene in the `/nekoyume/Asseta/_Scenes/`.

Reviewer: ```suggestion
- Double-click the Game scene in the `/nekoyume/Assets/_Scenes/`.
```

  <img
    alt="Unity Editor: `Game` scene"
    src="./images/unity-editor-game-scene.png"
    style="width:50vw" />

### 2. Click the `Play` button.

- Click the ▶️ button at the top and middle of the Unity editor.

  <img
    alt="Unity Editor: `Play` button"
    src="./images/unity-editor-play-button.png"
    style="width:50vw" />

### 3. Sign in.

> - Go to [3.1.](#31-sign-up) if there is no Protected Private Key in the Key Store path.

Reviewer: Discourse anchor warning lol

Author: I'll try on Discourse later.😁

- If there is no Protected Private Key in the Key Store path, go to [3.1.](#31-sign-up).

> - If you enter "PrivateKey" in the clo.json file, this process is skipped.

Reviewer: ```suggestion
- If you enter "PrivateKey" in the clo.json file, this process will be skipped.
```

[unity-install-with-hub-2021.3.5f1]: unityhub://2021.3.5f1/40eb3a945986
[unity-download-archive-page]: https://unity3d.com/get-unity/download/archive

# Clone the Nine Chronicles repository

여기서는 저장소를 복제하는 방법으로 [Git CLI][git-page]를 사용하는 방법과 [Fork][fork-page]를 사용하는 방법을 다룹니다.

[git-page]: https://git-scm.com/
[fork-page]: https://git-fork.com/

## Clone with Git CLI

### 1. 운영체제에 맞는 Git CLI를 설치합니다.

- [Windows](https://git-scm.com/download/windows)
- [Mac](https://git-scm.com/download/mac)
- [Linux](https://git-scm.com/download/linux)

### 2. 9C를 복제합니다.

```
$ git clone https://github.com/planetarium/NineChronicles.git
$ git submodule update --init --recursive
$ git config core.hooksPath hooks
```

### 3. main 브랜치로 체크아웃합니다.

```
$ git checkout main
$ git submodule update --recursive
```

## Clone with Fork

### 1. 운영체제에 맞는 Fork 설치합니다.

- [Windows(installer)](https://git-fork.com/update/win/ForkInstaller.exe)
- [Mac(2.20.2)](https://cdn.fork.dev/mac/Fork-2.20.2.dmg)
- 위 링크가 동작하지 않거나 다른 버전을 설치하고 싶다면 [Fork 페이지][fork-page]로 접속해서 다운로드 합니다.

### 2. Fork를 실행해서 저장소를 Clone 합니다.

- Fork 메뉴: File > Clone

  <img
    src="./images/Screen%20Shot%202022-08-30%20at%2011.48.24%20AM.png"

>     style="width:50vw" />

Reviewer: Do you have any reason not using MD style but HTML style? I know this works on markdown but I think we can use same format to embed images.
Or please add alt text at least.

Author: @U-lis I wanted to resize the image. OK I'll add alt text.

- URL과 Location을 입력합니다.
  - URL: `https://github.com/planetarium/NineChronicles.git`
  - Location: 아래의 이미지에서 저는 `/Users/seungmin/Repositories/NineChronicles`로 지정했습니다.

  <img

>     src="./images/Screen%20Shot%202022-08-30%20at%2012.02.42%20PM.png"

Reviewer: I think we have to set image naming convention.

Author: @U-lis I renamed image files.
